src/btree/bt_rebalance.c

```diff
@@ -0,0 +1,485 @@
+
+#define	WT_REBALANCE_PROGRESS_INTERVAL	100
+	uint64_t progress;			/* Progress counter */
+
+	WT_ITEM *tmp1;				/* Temporary buffers */
+	WT_ITEM *tmp2;
+} WT_REBALANCE_STUFF;
+
+/*
+ * __rebalance_discard --
+ *	Free the allocated information.
+ */
+static void
+__rebalance_discard(WT_SESSION_IMPL *session, WT_REBALANCE_STUFF *rs)
+{
+	while (rs->leaf_next > 0) {
+		--rs->leaf_next;
+		__wt_free_ref(
+		    session, rs->leaf[rs->leaf_next], rs->type, false);
+	}
+	__wt_free(session, rs->leaf);
+
+	while (rs->fl_next > 0) {
+		--rs->fl_next;
+		__wt_free(session, rs->fl[rs->fl_next].addr);
+	}
+	__wt_free(session, rs->fl);
+}
+
+/*
+ * __rebalance_leaf_append --
+ *	Add a new entry to the list of leaf pages.
+ */
+static int
+__rebalance_leaf_append(WT_SESSION_IMPL *session,
+    const uint8_t *key, size_t key_len, uint64_t recno,
+    const uint8_t *addr, size_t addr_len, u_int addr_type,
+    WT_REBALANCE_STUFF *rs)
+{
+	WT_ADDR *copy_addr;
+	WT_REF *copy;
+
+	WT_RET(__wt_verbose(session, WT_VERB_REBALANCE,
+	    "rebalance leaf-list append %s, %s",
+	    __wt_buf_set_printable(session, key, key_len, rs->tmp2),
+	    __wt_addr_string(session, addr, addr_len, rs->tmp1)));
+
+	/* Allocate and initialize a new leaf page reference. */
+	WT_RET(__wt_realloc_def(
+	    session, &rs->leaf_allocated, rs->leaf_next + 1, &rs->leaf));
+	WT_RET(__wt_calloc_one(session, &copy));
+	rs->leaf[rs->leaf_next++] = copy;
+
+	copy->page = NULL;
+	copy->home = NULL;
+	copy->pindex_hint = 0;
+	copy->state = WT_REF_DISK;
+
+	WT_RET(__wt_calloc_one(session, &copy_addr));
+	copy->addr = copy_addr;
+	WT_RET(__wt_strndup(session, addr, addr_len, &copy_addr->addr));
+	copy_addr->size = (uint8_t)addr_len;
+	copy_addr->type = (uint8_t)addr_type;
+
+	if (recno == WT_RECNO_OOB)
+		WT_RET(__wt_row_ikey(session, 0, key, key_len, copy));
+	else
+		copy->key.recno = recno;
+
+	copy->page_del = NULL;
+	return (0);
+}
+
+/*
+ * __rebalance_fl_append --
+ *	Add a new entry to the free list.
+ */
+static int
+__rebalance_fl_append(WT_SESSION_IMPL *session,
+    const uint8_t *addr, size_t addr_len, WT_REBALANCE_STUFF *rs)
+{
+	WT_ADDR *copy;
+
+	WT_RET(__wt_realloc_def(
+	    session, &rs->fl_allocated, rs->fl_next + 1, &rs->fl));
+	copy = &rs->fl[rs->fl_next++];
+
+	WT_RET(__wt_strndup(session, addr, addr_len, &copy->addr));
+	copy->size = (uint8_t)addr_len;
+	copy->type = 0;
+
+	return (0);
+}
+
+/*
+ * __rebalance_internal --
+ *	Build an in-memory page that references all of the leaf pages we've
+ * found.
+ */
+static int
+__rebalance_internal(WT_SESSION_IMPL *session, WT_REBALANCE_STUFF *rs)
+{
+	WT_BTREE *btree;
+	WT_DECL_RET;
+	WT_PAGE *page;
+	WT_PAGE_INDEX *pindex;
+	WT_REF **refp;
+	uint32_t i, leaf_next;
+
+	btree = S2BT(session);
+
+	/*
+	 * There's a limit to the number of pages we can rebalance: the number
+	 * of elements on a page is a 4B quantity and it's technically possible
+	 * there could be more pages than that in a tree.
+	 */
```

Comment:
@erbenmo, since `rebalance` is an exclusive operation, the tree will be flushed to disk when the operation completes.  That is independent of checkpoints or `WT_CONNECTION::close`.